File: apps/backend/ModuAgent/evolution/registry/rollback_mechanism.py
from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Dict, List, Optional, Tuple

if TYPE_CHECKING:
    from evolution.registry.versioned_store import VersionedComponentStore
    from core.registry import ComponentRegistry

logger = logging.getLogger(__name__)


class RollbackMechanism:
    """基于质量回退的自动回滚机制。"""

    def __init__(
        self,
        version_store: "VersionedComponentStore",
        registry: "ComponentRegistry",
        rollback_threshold: float = 0.7,
    ):
        self._version_store = version_store
        self._registry = registry
        self._rollback_threshold = rollback_threshold
        self._quality_records: Dict[str, List[Tuple[str, float]]] = {}
        self._rollback_count: int = 0

    def record_and_check(
        self,
        component_name: str,
        version: str,
        quality_score: float,
    ) -> bool:
        """记录质量得分并在需要时回滚。

        Returns:
            是否发生了回滚
        """
        if component_name not in self._quality_records:
            self._quality_records[component_name] = []
        self._quality_records[component_name].append((version, quality_score))

        logger.info(
            "Recorded quality score %.3f for %s version %s",
            quality_score,
            component_name,
            version,
        )

        if quality_score < self._rollback_threshold:
            logger.warning(
                "Quality score %.3f below threshold %.3f for %s",
                quality_score,
                self._rollback_threshold,
                component_name,
            )
            stable_version = self._find_stable_version(component_name)
            if stable_version:
                return self.rollback_to_version(component_name, stable_version)
            else:
                logger.error(
                    "No stable version found for %s to rollback to",
                    component_name,
                )
                return False

        return False

    def rollback_to_version(
        self,
        component_name: str,
        version: str,
    ) -> bool:
        """回滚到指定版本。

        流程：
        1. 从 version_store 获取版本快照
        2. 调用 registry.swap_component() 应用版本
        3. 记录回滚事件
        """
        snapshot = self._version_store.get_version(component_name, version)
        if snapshot is None:
            logger.error(
                "Failed to get snapshot for %s version %s",
                component_name,
                version,
            )
            return False

        category = snapshot.get("category")
        component = snapshot.get("component")

        if category is None or component is None:
            logger.error(
                "Invalid snapshot format for %s version %s",
                component_name,
                version,
            )
            return False

        success = self._registry.swap_component(category, component_name, component)
        if success:
            self._rollback_count += 1
            logger.info(
                "Successfully rolled back %s to version %s (rollback #%d)",
                component_name,
                version,
                self._rollback_count,
            )
        return success

    def _find_stable_version(self, component_name: str) -> Optional[str]:
        """找到满足质量阈值的最稳定版本。"""
        records = self._quality_records.get(component_name, [])
        for version, score in reversed(records):
            if score >= self._rollback_threshold:
                return version
        return None

    def get_quality_history(self, component_name: str) -> List[Tuple[str, float]]:
        """获取组件的质量历史记录。"""
        return self._quality_records.get(component_name, [])

    def get_rollback_count(self) -> int:
        """获取回滚总次数。"""
        return self._rollback_count

File: packages/modu-agent/src/evolution/rollback-mechanism.ts
// 对应 Python: evolution/registry/rollback_mechanism.py
// RollbackMechanism: 基于质量回退的自动回滚机制

import type { ComponentRegistry } from '../core/registry.js'
import type { VersionedComponentStore } from './versioned-store.js'

const logger = {
  info: (msg: string, ...args: any[]) => console.info(`[rollback] ${msg}`, ...args),
  warning: (msg: string, ...args: any[]) => console.warn(`[rollback] ${msg}`, ...args),
  error: (msg: string, ...args: any[]) => console.error(`[rollback] ${msg}`, ...args),
  debug: (msg: string, ...args: any[]) => console.debug(`[rollback] ${msg}`, ...args),
}

/** 版本与质量得分记录。 */
export type QualityRecord = [string, number]

/**
 * 基于质量回退的自动回滚机制。
 *
 * 对应 Python RollbackMechanism。
 * 依赖 VersionedComponentStore 的版本快照与 ComponentRegistry 的热替换能力。
 */
export class RollbackMechanism {
  private _versionStore: VersionedComponentStore
  private _registry: ComponentRegistry
  private _rollbackThreshold: number
  private _qualityRecords: Map<string, QualityRecord[]>
  private _rollbackCount: number

  constructor(
    versionStore: VersionedComponentStore,
    registry: ComponentRegistry,
    rollbackThreshold: number = 0.7,
  ) {
    this._versionStore = versionStore
    this._registry = registry
    this._rollbackThreshold = rollbackThreshold
    this._qualityRecords = new Map()
    this._rollbackCount = 0
  }

  /**
   * 记录质量得分并在需要时回滚。
   *
   * @returns 是否发生了回滚
   */
  recordAndCheck(
    componentName: string,
    version: string,
    qualityScore: number,
  ): boolean {
    if (!this._qualityRecords.has(componentName)) {
      this._qualityRecords.set(componentName, [])
    }
    this._qualityRecords.get(componentName)!.push([version, qualityScore])

    logger.info(
      'Recorded quality score %.3f for %s version %s',
      qualityScore,
      componentName,
      version,
    )

    if (qualityScore < this._rollbackThreshold) {
      logger.warning(
        'Quality score %.3f below threshold %.3f for %s',
        qualityScore,
        this._rollbackThreshold,
        componentName,
      )
      const stableVersion = this._findStableVersion(componentName)
      if (stableVersion) {
        return this.rollbackToVersion(componentName, stableVersion)
      } else {
        logger.error(
          'No stable version found for %s to rollback to',
          componentName,
        )
        return false
      }
    }

    return false
  }

  /**
   * 回滚到指定版本。
   *
   * 流程：
   * 1. 从 version_store 获取版本快照
   * 2. 调用 registry.swapComponent() 应用版本
   * 3. 记录回滚事件
   */
  rollbackToVersion(componentName: string, version: string): boolean {
    const snapshot = this._versionStore.getVersion(componentName, version)
    if (snapshot === null) {
      logger.error(
        'Failed to get snapshot for %s version %s',
        componentName,
        version,
      )
      return false
    }

    const category = snapshot['category']
    const component = snapshot['component']

    if (category === undefined || category === null || component === undefined || component === null) {
      logger.error(
        'Invalid snapshot format for %s version %s',
        componentName,
        version,
      )
      return false
    }

    const success = this._registry.swapComponent(category, componentName, component)
    if (success) {
      this._rollbackCount += 1
      logger.info(
        'Successfully rolled back %s to version %s (rollback #%d)',
        componentName,
        version,
        this._rollbackCount,
      )
    }
    return success
  }

  /** 找到满足质量阈值的最稳定版本。 */
  private _findStableVersion(componentName: string): string | null {
    const records = this._qualityRecords.get(componentName) ?? []
    // 从最近记录向前查找（对应 Python reversed(records)）
    for (let i = records.length - 1; i >= 0; i--) {
      const [version, score] = records[i]
      if (score >= this._rollbackThreshold) {
        return version
      }
    }
    return null
  }

  /** 获取组件的质量历史记录。 */
  getQualityHistory(componentName: string): QualityRecord[] {
    return this._qualityRecords.get(componentName) ?? []
  }

  /** 获取回滚总次数。 */
  getRollbackCount(): number {
    return this._rollbackCount
  }
}
